### ISSUE TITLE
Implement Keystone Application Credentials support

### ISSUE THREAD

**Issue by zioproto** [gophercloud/gophercloud#1214]
This is a feature request.

Keystone Application Credentials seems to be not yet supported by gophercloud:

https://docs.openstack.org/keystone/queens/user/application_credentials.html

Because these credentials are meant to be used with scripts and deployments tools, it would be very nice to have them supported in gophercloud.

Thank you

**Comment by zioproto** [comment]
Related github issues blocked by this missing feature:

- https://github.com/appscode/guard/issues/183
- https://github.com/kubernetes/cloud-provider-openstack/issues/232

**Comment by jtopjian** [comment]
Hi @zioproto !

> Related github issues blocked by this missing feature:

Just to give you a realistic expectation, the most likely way for this to get added is for someone to submit it. I've added a "help wanted" label to more easily identify it.

**Comment by zioproto** [comment]
@jtopjian does it make sense to copy the folder [tokens](https://github.com/gophercloud/gophercloud/tree/master/openstack/identity/v3/tokens) into a new folder called `application_credentials` to use the same code structure ? I would like to try to implement the basic functionality of authenticating with the application credential instead of the token. It is possible to add later the functionalities like create/delete of application credentials. If you can point me to the right direction on how to start I can try to implement this. Thank you

**Comment by jtopjian** [comment]
> does it make sense to copy the folder tokens into a new folder called application_credentials to use the same code structure ?

Off the top of my head, I'm not sure. I can review the application credential API later on today and let you know if that makes sense. However, if you're just looking for a template to use, I recommend reading the [contributor tutorial](https://github.com/gophercloud/gophercloud/tree/master/docs/contributor-tutorial) since it includes a template as well as a lot of other useful information on how to quickly get code merged.

Let me know if you need any help or have any questions. I'm at a conference this week, but I'll do my best to get back quickly 😄

**Comment by zioproto** [comment]
@jtopjian I have a question, but take your time to answer because I am leaving the office and I will be offline the next 16 hours ;) 

My 1st goal is to issue a user token, but instead of username and password I want to use a keystone application credential.

As explained here:
https://specs.openstack.org/openstack/keystone-specs/specs/keystone/backlog/application-credentials.html

The URL will be `POST /v3/auth/tokens`

and the example payload looks like:

```
    "auth": {
        "identity": {
            "methods": [
                "application_credential"
            ],
            "application_credential": {
                "name": "backup",
                "project": {
                    "id": "1a6f968a-cebe-4265-9b36-f3ca2801296c"
                },
                "secret": "a49670c3c18b9e079b9cfaf51634f563dc8ae3070db2..."
            }
        }
    }
}
```
Because this operation is about issuing a new token, it is correct that implementing this functionality is basically a patch in the [tokens](https://github.com/gophercloud/gophercloud/tree/master/openstack/identity/v3/tokens) folder ?

I understand that to create/delete application credentials the new API must be implemented, and then a completely new folder must be created. But I would like to start just with the token issue, so that if a user already has a application credential created with the openstack client, then he can use it with a gophercloud application to issue new tokens.

Does it make sense ? Thank you

**Comment by jtopjian** [comment]
> Because this operation is about issuing a new token, it is correct that implementing this functionality is basically a patch in the tokens folder ?

In every other case besides this one, yes 😄 You've stumbled upon a unique part of Gophercloud and I'll try to explain it as best as I can, but let me know if you have questions.

https://github.com/gophercloud/gophercloud/blob/master/openstack/identity/v3/tokens/requests.go is where all Token requests go. To build the payload body similar to what you provided, the [`ToTokenV3CreateMap`](https://github.com/gophercloud/gophercloud/blob/master/openstack/identity/v3/tokens/requests.go#L59) method is called. This method is unique to Gophercloud in that at [this](https://github.com/gophercloud/gophercloud/blob/master/openstack/identity/v3/tokens/requests.go#L70) line, it calls:

https://github.com/gophercloud/gophercloud/blob/master/auth_options.go#L130

to do all of the work. Notice how this `auth_options.go` file is at the root of Gophercloud. This is kind of like a bootstrapping process to authenticate the client but also retain the ability to create further tokens in a dedicated package.

Ideally, the best way to implement the ability to create a token using app credentials would be to add some new fields to:

* https://github.com/gophercloud/gophercloud/blob/master/openstack/identity/v3/tokens/requests.go#L24-L56
* https://github.com/gophercloud/gophercloud/blob/master/auth_options.go#L30-L87

and build the request body here: https://github.com/gophercloud/gophercloud/blob/master/auth_options.go#L130-L274

But be careful because you'll want to make sure the existing functionality is retained. 

This is do-able, but the implementation is unique to how most other things in Gophercloud are put together. Let me know if you have any questions or need any help.

**Comment by attardi** [comment]
The extension to AuthOptions could be this?

    // Authentication through Application Credentials requires supplying name, project and secret
    // For project we can use TenantID
    ApplicationCredentials   string `json:"-"`
    ApplicationCredentialsSecret   string `json:"-"`

and then:

    func (opts *AuthOptions) ToTokenV3CreateMap(scope map[string]interface{}) (map[string]interface{}, error) {
    	gophercloudAuthOpts := gophercloud.AuthOptions{
		Username:    opts.Username,
		UserID:      opts.UserID,
		Password:    opts.Password,
		DomainID:    opts.DomainID,
		DomainName:  opts.DomainName,
		AllowReauth: opts.AllowReauth,
		TokenID:     opts.TokenID,
 		TenantID:   opts.TenantID,
		ApplicationCredentials: opts.ApplicationCredentials,
		ApplicationCredentialsSecret: opts.ApplicationCredentialsSecret,
    	}
    	return gophercloudAuthOpts.ToTokenV3CreateMap(scope)
    }

and then add to https://github.com/gophercloud/gophercloud/blob/master/auth_options.go#L170

        type applicationCredentialsReq struct {
		Name     *string    `json:"name,omitempty"`
		ProjectID *string     `json:"projectID,omitempty"`
		Secret   *string `json:"secret,omitempty"`
	}

and to https://github.com/gophercloud/gophercloud/blob/master/auth_options.go#L191

	} else if opts.ApplicationCredentials != "" {
            // Configure the request for ApplicationCredentials authentication.
            req.Auth.Identity.Methods = []string{"applicationCredentials"}
            req.Auth.Identity.ApplicationCredentials = &applicationCredentialsReq{
            Name: &opts.ApplicationCredentials,
            ProjectID: &opts.TenantID,
            Secret: &opts.ApplicationCredentialsSecret,
            }
	} else {
            // If no password or token ID or ApplicationCredentials are available, authentication can't continue.
            return nil, ErrMissingPassword{}
        }

**Comment by jtopjian** [comment]
I think we should change `opts.ApplicationCredentials` to `opts.ApplicationCredentialsName`, though, just to be explicit about what the field is.

Other than that, it looks like it'll work. I recommend trying it out with some unit tests. The OpenLab testing should be able to verify if original functionality still works. I can double-check it manually against some clouds using various auth combinations, too.

**Comment by attardi** [comment]
Should the ApplicatinCredentialReq struct correspond to the payload to be submitted?
For a payload like

    "application_credential": {
            "name": "backup",
            "project": {
                  "id": "1a6f968a-cebe-4265-9b36-f3ca2801296c"
            },
            "secret": "a49670c3c18b9e079b9cfaf51634f563dc8ae3070db2..."
        }

then it should be:

    type applicationCredentialReq struct {
	Name     *string    `json:"name,omitempty"`
	Project   *projectReq     `json:"project,omitempty"`
	Secret   *string `json:"secret,omitempty"`
    }

and then create the request as:

        req.Auth.Identity.Methods = []string{"application_credential"}
        req.Auth.Identity.ApplicationCredential = &applicationCredentialReq{
        Name: &opts.ApplicationCredentialName,
        Project: &projectReq{ID: &opts.TenantID}
        Secret: &opts.ApplicationCredentialSecret,

It is necessary also to extend the identityReq struct 

    type identityReq struct {
		Methods  []string     `json:"methods"`
		Password *passwordReq `json:"password,omitempty"`
		Token    *tokenReq    `json:"token,omitempty"`
		ApplicationCredential    *applicationCredentialReq    `json:"application_credential,omitempty"`
	}

Finally in AuthOptions

    // Authentication through Application Credentials requires supplying name, project and secret
    // For project we can use TenantID
    ApplicationCredentialName   string `json:"-"`
    ApplicationCredentialSecret   string `json:"-"`

**Comment by jtopjian** [comment]
> Should the ApplicatinCredentialReq struct correspond to the payload to be submitted?

I don't think so. Since we can safely re-use `TenantID` as the Project ID in the `application_credential` section, that negates a need to have a dedicated `ApplicationCredentialReq` structure.

**Comment by jtopjian** [comment]
Oh, you meant creating the `applicationCredentialReq` struct within the `ToTokenV3CreateMap` method. In that case, yes. I thought you had meant adding a `applicationCredentialReq` struct as a field to the `AuthOptions` struct.

**Comment by zioproto** [comment]
@jtopjian I will also patch also https://github.com/gophercloud/gophercloud/blob/master/openstack/auth_env.go#L32 to read `OS_APPLICATION_CREDENTIAL_NAME` and `OS_APPLICATION_CREDENTIAL_SECRET` env var names.
I dont find any other place in the code where to read the auth options, right ?

**Comment by jtopjian** [comment]
> I dont find any other place in the code where to read the auth options, right ?

I think you've spotted everything :)

#1224 is looking good. It passed the OpenLab tests which is a great sign that the changes didn't break user/pass authentication.

We'll need some unit tests added [here](https://github.com/gophercloud/gophercloud/blob/master/openstack/identity/v3/tokens/testing/requests_test.go). After that, it should be ready for review.

**Comment by zioproto** [comment]
I dumped a request from the official `openstack` client using this fake keystone server on localhost https://gist.github.com/zioproto/ac42e43da953e156d682b8d6936dd9e6

The request looks like:
```
{
    "auth": {
        "identity": {
            "application_credential": {
                "name": "test09",
                "secret": "mysecret",
                "user": {
                    "domain": {
                        "name": "cloudusers"
                    },
                    "name": "saverio.proto@switch.ch"
                }
            },
            "methods": [
                "application_credential"
            ]
        }
    }
}
```

It looks like `applicationCredentialReq` needs the user information and not the project information. This is most probably an error in this documentation page https://specs.openstack.org/openstack/keystone-specs/specs/keystone/backlog/application-credentials.html 

I updated the PR to reflect this change.

**Comment by zioproto** [comment]
I got confirmation that using project ID is a documentation error.
Correct requests JSON templates can be found here.
https://developer.openstack.org/api-ref/identity/v3/index.html#application-credentials

**Comment by zioproto** [comment]
More than a documentation error in keystone, it turns out that Google gives bad results, and proposes on top of search results outdated documentation from the `backlog` folder.

The correct specs are published here https://specs.openstack.org/openstack/keystone-specs/specs/keystone/queens/application-credentials.html

**Comment by jtopjian** [comment]
@zioproto This is why we require references to the actual OpenStack service code to validate a feature's implementation.

In #1224 you have provided https://github.com/openstack/keystone/blob/589152d094b248da81dc88db2449fb560985ae8b/keystone/conf/constants.py#L20, but we're going to need to see more than that in order to fully validate the implementation. I would suggest searching https://github.com/openstack/keystone for `application_credentials`, finding where Keystone receives an application credential body and parses it.

**Comment by jtopjian** [comment]
Implemented in #1224
